addons/portal/static/src/js/portal_composer.js

```diff
@@ -138,15 +137,42 @@ var PortalComposer = publicWidget.Widget.extend({
         });
     },
     /**
-     * Returns a Promise that is never resolved to prevent sending the form
-     * twice when clicking twice on the button, in combination with the `async`
-     * in the event definition.
-     *
+     * prepares data to send message
      * @private
+     */
+    _prepareMessageData: function () {
+        return _.extend(this.options || {}, {
+            'message': this.$('textarea[name="message"]').val(),
+            'attachment_ids': _.pluck(this.attachments, 'id'),
+            'attachment_tokens': _.pluck(this.attachments, 'access_token'),
+        });
+    },
+    /**
+     * Send message using rpc call and display new message and message count
+     * @private
+     * @param {Event} ev
      * @returns {Promise}
      */
-    _onSubmitButtonClick: function () {
-        return new Promise(function (resolve, reject) {});
+    _onSubmitButtonClick: function (ev) {
+        ev.preventDefault();
+        return new Promise((resolve, reject) => {
+            if (!this.$inputTextarea.val().trim() && !this.attachments.length) {
+                this.$inputTextarea.addClass('border-danger');
+                let text = _t('Oops! Some fields are required. Please make sure that message or attachment is not empty!');
+                this.$(".o_portal_chatter_composer_error").html(text).removeClass('d-none');
+                return reject(false);
+            }
+            this._rpc({
+                route: ev.currentTarget.getAttribute('data-action'),
+                params: this._prepareMessageData(),
+            }).then((result) => {
+                core.bus.trigger('reload_chatter_content', result);
+                resolve(result);
+            }).guardedCatch((error) => {
+                console.error("Oops! Something went wrong while posting message.", error);
```

Comment:
This is actually something we do very incorrectly in general.

`guardedCatch` is an aberration that we had to introduce because we often used "reject" as an "else" in our code, whereas "reject" is actually alike to "throw". It's like using exceptions for doing conditions.

Ideally `catch` should be used to handle real failure cases (such as the RPC actually failing), it's something we currently don't do often, and it can lead to real issues being hidden and the user having no clue of what happened (or what didn't happen). This has been improved a bit in v13 with the error manager. But we still don't handle issues contextually, in the part of the interface where it happened.

That said here a `console.error` isn't going to help the user either, so in that case I would remove it indeed. This is only useful if the interface is adapted accordingly to tell the user what happened and what to do.

Reply:
Also note that here a promise is used inside the constructor of another promise, which is an anti-pattern to avoid: since the catch of the rpc is not defined, if there is any real issue, the catch of the top-level promise will not be called.

Instead promise chaining should be used:
- no promise constructor
- `return rpc[...].then[...].guardedCatch[...];`

And the check of required fields is synchronous and should not be wrapped inside a promise at all.